XS✔ ◾ Sql server performance (#7655)
SSWConsulting/SSW.Rules.Content: rules/sql-server-cpu-pressure/rule.md

---
type: rule
archivedreason: 
title: What to do about SQL Server CPU Pressure?
guid: 870f4673-3a45-48d9-b73d-14921a77bc70
uri: sql-performance-cpu-pressure
created: 2023-12-27T07:26:54.0000000Z
authors:
  - title: Bryden Oliver
    url: https://ssw.com.au/people/bryden-oliver
related: 
  - identify-sql-performance-sql-server
  - identify-sql-performance-azure
  - sql-server-memory-pressure
  - sql-server-io-pressure
---
So you've identified that your SQL Server is under CPU pressure. What can you do about it?

<!--endintro-->

Here's a simple in-depth presentation covering things that can help reduce CPU pressure.

> 1. Should you throw hardware at the problem

Reviewer: **issue**: Grammar - the first point is written as a question (without question mark), while the remaining points are actions.
```suggestion
1. Throw hardware at the problem
```

           COALESCE(QUOTENAME(DB_NAME(st.dbid)) + N'.' + QUOTENAME(OBJECT_SCHEMA_NAME(st.objectid, st.dbid)) 
           + N'.' + QUOTENAME(OBJECT_NAME(st.objectid, st.dbid)), '') AS command_text,
           r.command,
           s.login_name,
           s.host_name,
           s.program_name,
           s.last_request_end_time,
           s.login_time,
           r.open_transaction_count
FROM sys.dm_exec_sessions AS s
JOIN sys.dm_exec_requests AS r ON r.session_id = s.session_id CROSS APPLY sys.Dm_exec_sql_text(r.sql_handle) AS st
WHERE r.session_id != @@SPID
ORDER BY r.cpu_time DESC
```

If queries aren't driving the CPU currently, try the following query.

``` sql
SELECT TOP 10 st.text AS batch_text,
    SUBSTRING(st.TEXT, (qs.statement_start_offset / 2) + 1, ((CASE qs.statement_end_offset WHEN - 1 THEN DATALENGTH(st.TEXT) ELSE qs.statement_end_offset END - qs.statement_start_offset) / 2) + 1) AS statement_text,
    (qs.total_worker_time / 1000) / qs.execution_count AS avg_cpu_time_ms,
    (qs.total_elapsed_time / 1000) / qs.execution_count AS avg_elapsed_time_ms,
    qs.total_logical_reads / qs.execution_count AS avg_logical_reads,

>     (qs.total_worker_time / 1000) AS cumulative_cpu_time_all_executions_ms,

Reviewer: **nitpick**: Formatting consistency - are the parentheses required (when they weren't on the line above)?

Author: Fixed

    (qs.total_elapsed_time / 1000) AS cumulative_elapsed_time_all_executions_ms
FROM sys.dm_exec_query_stats qs
CROSS APPLY sys.dm_exec_sql_text(sql_handle) st

> ORDER BY(qs.total_worker_time / qs.execution_count) DESC

Reviewer: **nitpick**: Spacing.
```suggestion
ORDER BY (qs.total_worker_time / qs.execution_count) DESC
```

```

## Identify missing indexes

> Indexes can dramatically improve query performance. SQL Server has inbuilt mechanisms to try and identify indexes that would aid a particular query. Running the following sql idenitfies the top 50 queries that SQL Server has identified a potential missing index.

Reviewer: **question**: Is the ranking of the queries based on how frequently they have been called, or an assessment of the indexes on the performance?
**issue**: Spelling and grammar.
```suggestion
Indexes can dramatically improve query performance. SQL Server has inbuilt mechanisms to try and identify indexes that would aid a particular query. Running the following SQL lists the top 50 queries that SQL Server has identified as potentially missing indexes.
```

``` sql
SELECT
    qs_cpu.total_worker_time / 1000 AS total_cpu_time_ms,
    q.[text],
    p.query_plan,
    qs_cpu.execution_count,
    q.dbid,
    q.objectid,
    q.encrypted AS text_encrypted
FROM
    (SELECT TOP 50 qs.plan_handle,
     qs.total_worker_time,
     qs.execution_count FROM sys.dm_exec_query_stats qs ORDER BY qs.total_worker_time DESC) AS qs_cpu
CROSS APPLY sys.dm_exec_sql_text(plan_handle) AS q
CROSS APPLY sys.dm_exec_query_plan(plan_handle) p
WHERE p.query_plan.exist('declare namespace 
        qplan = "http://schemas.microsoft.com/sqlserver/2004/07/showplan";
        //qplan:MissingIndexes')=1
```

This is super useful for giving suggestions. Otherwise, you may need to manually identify potential indexes and test those.

`youtube: https://youtu.be/l18ltcOVN4I?si=Ejf0sGKECfuqFw2L`

## Identifying parameter-sensitive problems

Try running

``` sql
DBCC FREEPROCCACHE

> ```

Reviewer: **question**: What is DBCC?

Author: Added a description

           s.last_request_end_time,
           s.login_time,
           r.open_transaction_count
FROM sys.dm_exec_sessions AS s
JOIN sys.dm_exec_requests AS r ON r.session_id = s.session_id CROSS APPLY sys.Dm_exec_sql_text(r.sql_handle) AS st
WHERE r.session_id != @@SPID
ORDER BY r.cpu_time DESC
```

If queries aren't driving the CPU currently, try the following query.

``` sql
SELECT TOP 10 st.text AS batch_text,
    SUBSTRING(st.TEXT, (qs.statement_start_offset / 2) + 1, ((CASE qs.statement_end_offset WHEN - 1 THEN DATALENGTH(st.TEXT) ELSE qs.statement_end_offset END - qs.statement_start_offset) / 2) + 1) AS statement_text,
    (qs.total_worker_time / 1000) / qs.execution_count AS avg_cpu_time_ms,
    (qs.total_elapsed_time / 1000) / qs.execution_count AS avg_elapsed_time_ms,
    (qs.total_logical_reads / qs.execution_count) AS avg_logical_reads,
    (qs.total_worker_time / 1000) AS cumulative_cpu_time_all_executions_ms,
    (qs.total_elapsed_time / 1000) AS cumulative_elapsed_time_all_executions_ms
FROM sys.dm_exec_query_stats qs
CROSS APPLY sys.dm_exec_sql_text(sql_handle) st
ORDER BY (qs.total_worker_time / qs.execution_count) DESC
```

## Identify missing indexes

> Indexes can dramatically improve query performance. SQL Server has inbuilt mechanisms to try and identify indexes that would aid a particular query. Running the following sql identifies the 50 queries consuming the most CPU where SQL Server has identified that there is potentially a missing index.

Reviewer: **nitpick**: Consistent capitalization.
```suggestion
Indexes can dramatically improve query performance. SQL Server has inbuilt mechanisms to try and identify indexes that would aid a particular query. Running the following SQL identifies the 50 queries consuming the most CPU where SQL Server has identified that there is potentially a missing index.
```
